### PR TITLE
Add API server, Client and CLI

### PR DESCRIPTION
**Pull request by lxea**
This adds the api server as described in the DESIGN.md. It also adds a Client which is used in a test where the server is actually run, and in addition it adds the user CLI

### CODE REVIEW

File: api/common.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package api
+
+import "github.com/google/uuid"
+
+// common holds types that can be common to both a client and a server
+
+type StartReq struct {
```
Comment (russjones) [inline api/common.go:7]: Add comments to public structs and fields.

File: api/api.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+func (h *handlers) authnMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		user, pass, ok := r.BasicAuth()
+		if !ok {
+			jsonError(w, "Unauthenticated", http.StatusForbidden)
+			return
+		}
+		token := h.users[user]
+		if token != pass {
```
Comment (russjones) [inline api/api.go:101]: Use https://pkg.go.dev/crypto/subtle#ConstantTimeCompare to avoid timing attacks.

File: api/api.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+			jsonError(w, "Invalid Method", http.StatusMethodNotAllowed)
+			return
+		}
+
+		next.ServeHTTP(w, r)
+	})
+}
+
+func (h *handlers) handleStart(w http.ResponseWriter, r *http.Request) {
+	user, _, _ := r.BasicAuth()
```
Comment (russjones) [inline api/api.go:147]: Minor suggestion, you do not have to make this change, but you can attach this information to the `*http.Request` in the handlers and just propagate it along to all subsequent handlers. That way you don't have to re-parse headers in each handler.

https://pkg.go.dev/context#WithValue

File: api/api.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+		return
+	}
+	h.procLock.Lock()
+	userProcs, ok := h.userProcs[user]
+	if !ok {
+		userProcs = map[uuid.UUID]struct{}{}
+		h.userProcs[user] = userProcs
+	}
+	userProcs[id] = struct{}{}
+	h.procLock.Unlock()
```
Comment (russjones) [inline api/api.go:167]: I'd pull this out into it's own function, something like `getOrCreate*`, plus this allows you to use the standard lock and defer unlock pattern.

File: api/api.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+	stream := r.URL.Query().Get("stream")
+	if stream == "" {
+		stream = "stdout"
+	}
+	if !executor.OutStream(stream).Valid() {
+		jsonError(w, "Invalid stream type", http.StatusBadRequest)
+		return
+	}
+	logPath := filepath.Join(h.logDir, qID, stream)
+	f, err := os.Open(logPath)
```
Comment (russjones) [inline api/api.go:220]: Don't forget to close this, otherwise you'll leak file descriptors: `defer f.Close()`.

File: api/api.go [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+	qID := r.URL.Query().Get("id")
+	stream := r.URL.Query().Get("stream")
+	if stream == "" {
+		stream = "stdout"
+	}
+	if !executor.OutStream(stream).Valid() {
+		jsonError(w, "Invalid stream type", http.StatusBadRequest)
+		return
+	}
+	logPath := filepath.Join(h.logDir, qID, stream)
```
Comment (russjones) [inline api/api.go:219]: Is this vulnerable to path traversal attacks?

https://dzx.cz/2021/04/02/go_path_traversal/

Can the client pass in `../../../../../etc/passwd` and read `/etc/passwd`?

File: api/api.go [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+func (h *handlers) authnMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		user, pass, ok := r.BasicAuth()
+		if !ok {
+			jsonError(w, "Unauthenticated", http.StatusForbidden)
+			return
+		}
+		token := h.users[user]
```
Comment (russjones) [inline api/api.go:100]: I think a data race exists here since multiple clients can be in this handler at the same time.

File: api/api.go (hunk 6 above)
Comment (lxea) [reply api/api.go:219]: This shouldnt be able to happen as the authzMiddleware validates that the id is a correctly formatted UUID, I wasnt sure if it was preferable to check this multiple times or not

File: api/api.go [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+	userProcs, ok := h.userProcs[user]
+	if !ok {
+		userProcs = map[uuid.UUID]struct{}{}
+		h.userProcs[user] = userProcs
+	}
+	return userProcs
+}
+
+func (h *handlers) handleStart(w http.ResponseWriter, r *http.Request) {
+	user := r.Context().Value(userCtxID).(string)
```
[the remark this answers is not in the document]
Comment (lxea) [reply api/api.go]: I wasnt sure, I've added a log small log message to make it a bit clearer

File: api/api.go [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+		if r.Method != method {
+			jsonError(w, "Invalid Method", http.StatusMethodNotAllowed)
+			return
+		}
+
+		next.ServeHTTP(w, r)
+	})
+}
+
+func (h *handlers) getOrCreateUserProcesses(user string, id uuid.UUID) map[uuid.UUID]struct{} {
```
Comment (timothyb89) [inline api/api.go:162]: Looks like `id` is unused?

File: api/api.go [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+		return
+	}
+	id, err := h.executor.Start(startOpts.Command, startOpts.Args...)
+	if err != nil {
+		jsonError(w, "Failed to execute command", http.StatusInternalServerError)
+		log.Println(err)
+		return
+	}
+	userProcs := h.getOrCreateUserProcesses(user, id)
+	userProcs[id] = struct{}{}
```
Comment (timothyb89) [inline api/api.go:193]: Hmm, it seems a bit problematic to write to this without holding the `procLock`

File: cmd/apiclient/main.go [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+	case "start":
+		start(client, args[1:])
+	case "stop":
+		stop(client, args[1:])
+	case "status":
+		status(client, args[1:])
+	case "output":
+		output(client, args[1:])
+	default:
+		fmt.Println("please provide one of start|stop|status|output options")
```
Comment (timothyb89) [inline cmd/apiclient/main.go:54]: Ideally this should return non-0 for the failure case

File: api/api.go [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+	// remove the default go support for tls1.2 and only support 1.3
+	// for some added security.
+	return http.ListenAndServeTLS(":8443", config.CrtPath, config.KeyPath, nil)
+}
+
+type handlers struct {
+	logDir    string
+	users     map[string]string
+	userLock  sync.RWMutex
+	userProcs map[string]map[uuid.UUID]struct{}
```
[the remark this answers is not in the document]
Comment (lxea) [reply api/api.go:89]: As it stands, if a user is creating multiple processes concurrently, all of the processes will get started but they will get synchronized before the id gets returned to the user. The global user lock needs to happen so as if a user has never created a process before tries to create multiple concurrent processes a race doesnt happen where one of the processes could get lost.

I've changed it now so the process set for a user can also be separately locked and then the write becomes this:
```go
func (h *handlers) updateOrCreateUserProcesses(user string, id uuid.UUID) {
	h.procLock.Lock() // stops the concurrent start race
	userProcs, ok := h.userProcs[user]
	if !ok {
		userProcs = &procs{p: map[uuid.UUID]struct{}{}}
		h.userProcs[user] = userProcs
	}
	h.procLock.Unlock()

	userProcs.Lock()
	defer userProcs.Unlock()
	userProcs.p[id] = struct{}{}
}
```
I'm not sure if this is better or worse really, or if there is a better way. If we consider the actual list of users to be constant (as they are in this case, but would not be in production) all of them could be initialized at startup and then the above could become something like this:

```go
func (h *handlers) updateOrCreateUserProcesses(user string, id uuid.UUID) {
	h.procLock.RLock() // stops the concurrent start race
	userProcs := h.userProcs[user]
	h.procLock.RUnlock()
	userProcs.Lock()
	defer userProcs.Unlock()
	userProcs.p[id] = struct{}{}
}
```

**Review by tcsc — APPROVED** [review]

**Review by timothyb89 — APPROVED** [review]

**Pull request merged**